[Material] Create an InkSparkle that matches the Material 3 ripple effect

clocksmith opened the pull request:
- Create an InkSparkle effect that matches the Material 3 ripple on Android 12. Similar to InkRipple, but uses the new custom shader API.
- Uses a generated SPIR-V Fragment Shader from a GLSL source.
- Uses a generated FragmentShaderManager to make uniform updates and compilation more legible.

Android Samples:
![black](https://user-images.githubusercontent.com/591699/157056479-4f4df278-23ad-469e-b7d1-472f1e05c516.gif)
![pink](https://user-images.githubusercontent.com/591699/157056488-4e0f70a3-4bbe-4762-ba1f-697ce2887b16.gif)
![green](https://user-images.githubusercontent.com/591699/157056500-db3e87ac-40d4-45c7-95b4-89cf697006fa.gif)

closes https://github.com/flutter/flutter/issues/82850

**Note on performance:**

Currently, the shader needs to be compiled at run time before the first InkSparkle can be painted. This takes 3-4 milliseconds on a Pixel 6 Pro and 30-35 milliseconds on a 2014 Moto E. The Moto E runs InkRipple at about 5 fps and the InkSparkle at about 5 fps. Even in the worst case, the async shader compilation is negligible, but it could be moved to MaterialApp or Theme creation instead.

Up next:
- https://github.com/flutter/flutter/issues/73163
- https://github.com/flutter/flutter/issues/98669
- https://github.com/flutter/flutter/issues/98670
- https://github.com/flutter/flutter/issues/98671

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

LasseRosenow commented:
In the gif above, I can see that the whole background immediately gets another color on touch, but as far as I know, the background should only change its color by the circular animation

clocksmith commented:
> In the gif above, I can see that the whole background immediately gets another color on touch, but as far as I know, the background should only change its color by the circular animation

@LasseRosenow Good catch, that is the highlight for mouse, need to verify that it does not show on touch only

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+  return math.max(d1, d2) / 2.0;
+}
+
+  /// Begin an m3 sparkle ripple, centered at [position] relative to
```
can you add documentation for all parameters and a short code sample showing usage?

nt4f04uNd replied:
Also, should it be "a Material 3"?

clocksmith replied:
Done

clocksmith replied:
Since a proper sample relies on dartpad, I added instructions, and will add a proper example once https://github.com/flutter/flutter/issues/85238 is resolved

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+    // This method is only called on press up.
+    _timeUpMillis = _time.inMilliseconds;
+    // TODO(clocksmith): Verify value for ripple settle time.
+    Timer(const Duration(milliseconds: 150), () {
```
I think it's convention to store constants at the top

clocksmith replied:
moved to top of class as static consts per flutter style guide

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+  final ShapeBorder? _customBorder;
+  final double _targetRadius;
+  final RectCallback? _clipCallback;
+  final TextDirection _textDirection;
```
I'd make the constructor arguments public and separate them from the private stuff like `_textDirection`

clocksmith replied:
Why public constructor args? This is modeled off ink_ripple.dart.

Can address this in https://github.com/flutter/flutter/issues/73163

clocksmith replied:
Resolving for now, reopen if you feel strongly about them being public

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+
+  /// Clips the canvas for an ink feature to be painted on the [canvas].
+  ///
+  /// This should be called before painting ink features that do not use
```
nit: reference [paintFeature]

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+  /// Clipping can happen in 3 different ways:
+  ///  1. If [customBorder] is provided, it is used to determine the path
+  ///     for clipping.
+  ///  2. If [customBorder] is null, and [borderRadius] is provided, the canvas
```
```suggestion
  ///  2. If a non-zero [borderRadius] is provided, the canvas
```

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+  ///     for clipping.
+  ///  2. If [customBorder] is null, and [borderRadius] is provided, the canvas
+  ///     is clipped by an [RRect] created from [clipCallback] and [borderRadius].
+  ///  3. If [borderRadius] is the default [BorderRadius.zero], then the [Rect] provided
```
```suggestion
  ///  3. Otherwise, it is clipped with the [Rect] provided by [clipCallback]
```

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+        uRippleDurationMs: 3000, // TODO(clocksmith): Confrim value
+        uPauseInflection: 0.185,
+        uPostPauseTimeScale: 1.5,
+        uPixelDensity: 1.0,
```
these should probably be consts defined at the top with explanatory comments

clocksmith replied:
This changed since last comment. There a quite a few double values from Android 12 that do not have explanations. They were tuned by designers to make the ripple look better. Added references to those Android 12 files.

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+  ByteBuffer get spirvByteBuffer => Uint8List.fromList(_spirvByteList).buffer;
+}
+
+const _glslString = '''
```
curious if there it would be "better" to have this be in a .glsl file that is read

clocksmith replied:
AH good question! This is not actually even necessary. It was generated as part of the generator library as a default behavior, but it is not actually needed for the InkSparkle, only the SPIR-V. Will move as a comment in the generated file instead.

guidezpl commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,765 @@
+      customBorder: customBorder,
+      radius: radius,
+      onRemoved: onRemoved,
+      textDirection: textDirection,
```
nit: move this line up to keep same order

engine-flutter-autoroll committed:
Roll Plugins from 6b0d4549eaa5 to c48a5f87757d (8 revisions) (#98504)

engine-flutter-autoroll committed:
Roll Engine from e683e0cd3be2 to 1b6755ce759a (3 revisions) (#98509)

christopherfujino committed:
[flutter_tools] do not validate maven upstream if cloud storage override provided (#98293)

engine-flutter-autoroll committed:
Roll Engine from 1b6755ce759a to 4463cf40fed9 (1 revision) (#98519)

Commit:
Roll Engine from 4463cf40fed9 to 5f0271a38988 + Fix test (#98479)

HansMuller committed:
Added a NavgationBar example with nested Navigators (#98440)

goderbauer committed:
Revert "[performance] Process dirty nodes from top to bottom during paint to avoid unnecessary layer tree walks" (#98520)

Renzo-Olivares committed:
Hide toolbar when selection is out of view (#98152)

* Hide toolbar when selection is out of view

* properly dispose of toolbar visibility listener

* Add test

* rename toolbarvisibility

* Make visibility for toolbar nullable

* Properly dispose of toolbar visibility listener

* Merge visibility methods into one

* properly dispose of start selection view listener

* Add some docs

* remove unnecessary null check

* more docs

* Update dispose order

Co-authored-by: Renzo Olivares <roliv@google.com>

engine-flutter-autoroll committed:
Roll Engine from 5f0271a38988 to 12609b45685b (1 revision) (#98528)

engine-flutter-autoroll committed:
Roll Engine from 12609b45685b to cbc6747fb7b3 (1 revision) (#98535)

engine-flutter-autoroll committed:
Roll Engine from cbc6747fb7b3 to b5ce63a145b5 (1 revision) (#98544)

chunhtai committed:
Add explanation to ChangeNotifier (#98295)

engine-flutter-autoroll committed:
Roll Engine from b5ce63a145b5 to 4ea2394b8df6 (2 revisions) (#98554)

fluttergithubbot committed:
Marks Mac_ios flutter_view_ios__start_up to be flaky (#98420)

TahaTesser committed:
`CupertinoPicker`: Update example (#98525)

engine-flutter-autoroll committed:
Roll Engine from 4ea2394b8df6 to b09cf532697b (2 revisions) (#98558)

markusaksli-nc committed:
Dismiss text selection toolbar with ESC (#98511)

engine-flutter-autoroll committed:
Roll Plugins from c48a5f87757d to 136f49b3e8ec (9 revisions) (#98586)

gaaclarke committed:
Added performance benchmarks to background platform channels for iOS. (#97991)

markusaksli-nc committed:
Dismiss Autocomplete with ESC (#97790)

markusaksli-nc committed:
Revert "Dismiss text selection toolbar with ESC (#98511)" (#98600)

This reverts commit 9407700aa4a843885d7acca324773f85e3766743.

TahaTesser committed:
Dismiss Modal Barrier on `handleTapCancel`  (#98191)

jmagman committed:
Dump backtrace when cannot attach to observatory (#98550)

mateusfccp committed:
Remove unused parameter and consequently unused variable (#98428)

GaryQian committed:
Specify android version for ci.yaml dependency (#98571)

engine-flutter-autoroll committed:
Roll Engine from b09cf532697b to 2cd5e113e84a (12 revisions) (#98606)

engine-flutter-autoroll committed:
Roll Engine from 2cd5e113e84a to a62228fc4cde (3 revisions) (#98625)

goderbauer committed:
Update example code and docs for InteractiveViewer.builder (#98623)

Piinks committed:
Remove deprecated RectangularSliderTrackShape.disabledThumbGapWidth (#98613)

engine-flutter-autoroll committed:
Roll Engine from a62228fc4cde to 57d4a350dea8 (1 revision) (#98630)

ditman committed:
[devicelab] Web benchmarks now run on Chromium 89+ (#98629)

engine-flutter-autoroll committed:
Roll Engine from 57d4a350dea8 to 754478998b5b (1 revision) (#98634)

GaryQian committed:
Upgrade android sdk for ci.yaml tests to 31 (#98638)

engine-flutter-autoroll committed:
Roll Engine from 754478998b5b to 7d33de0a372c (1 revision) (#98646)

skia-gold commented:
Gold has detected about 15 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 17 new digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/96598

engine-flutter-autoroll committed:
Roll Plugins from 136f49b3e8ec to ed29396208cd (12 revisions) (#98666)

engine-flutter-autoroll committed:
Roll Engine from 7d33de0a372c to 345a04b11f96 (1 revision) (#98678)

albertodev01 committed:
fix(FocusScope): Example fix (#98668)

engine-flutter-autoroll committed:
Roll Engine from 345a04b11f96 to 01ac6cf661a7 (4 revisions) (#98690)

engine-flutter-autoroll committed:
Roll Engine from 01ac6cf661a7 to 0ca8e1a24c81 (2 revisions) (#98698)

dependabot[bot] committed:
Bump github/codeql-action from 1.1.0 to 1.1.2 (#98699)

Piinks committed:
Update stretching overscroll clip behavior (#97678)

Piinks committed:
Remove deprecated UpdateLiveRegionEvent (#98615)

Commit:
Remove "beta-quality" for Windows (#98614)

engine-flutter-autoroll committed:
Roll Engine from 0ca8e1a24c81 to aa07e8f9304d (4 revisions) (#98708)

engine-flutter-autoroll committed:
Roll Engine from aa07e8f9304d to c5e89cbb7377 (6 revisions) (#98713)

engine-flutter-autoroll committed:
Roll Engine from c5e89cbb7377 to 605454e20143 (3 revisions) (#98718)

engine-flutter-autoroll committed:
Roll Engine from 605454e20143 to 0fd7713f2b5f (3 revisions) (#98727)

engine-flutter-autoroll committed:
Roll Engine from 0fd7713f2b5f to 9ef819b83b2b (1 revision) (#98734)

TahaTesser committed:
Remove `clipBehavior == Clip.none` conditions (#98503)

engine-flutter-autoroll committed:
Roll Engine from 9ef819b83b2b to c8d15c314135 (1 revision) (#98736)

TahaTesser committed:
Show `RefreshIndicator` on top when scroll's axis direction is up (matching native behaviour) (#93779)

engine-flutter-autoroll committed:
Roll Plugins from ed29396208cd to 1e7af5e1d66b (6 revisions) (#98751)

engine-flutter-autoroll committed:
Roll Engine from c8d15c314135 to 7c9b1d621dc8 (5 revisions) (#98754)

christopherfujino committed:
[flutter_tools] remove flaky web vm service test (#98540)

bkonyi committed:
Update dependencies to pull in new vm_service, dds, dwds, coverage, etc (#98513)

bkonyi committed:
Reland "Enable caching of CPU samples collected at application startup (#89600)" (#98769)

This reverts commit ac3c44355ae035b1e7f67881d956bcdfb8223e6f.

Piinks committed:
Remove deprecated VelocityTracker constructor (#98541)

Commit:
Improve Gradle retry logic (#96554)

engine-flutter-autoroll committed:
Roll Engine from 7c9b1d621dc8 to bad558812de9 (4 revisions) (#98768)

dependabot[bot] committed:
Bump ossf/scorecard-action from 1.0.3 to 1.0.4 (#98771)

christopherfujino committed:
[flutter_conductor] fix auto-tagging branch point (#98618)

christopherfujino committed:
[flutter_tools] deprecate the dev branch from the feature system (#98689)

goderbauer committed:
Add more tests to slider to avoid future breakages (#98772)

zanderso committed:
Revert "Add more tests to slider to avoid future breakages (#98772)" (#98783)

This reverts commit baa2803fe1da00b18c47797abc31db230a64f5df.

engine-flutter-autoroll committed:
Roll Engine from bad558812de9 to e0d70c0bc2c0 (2 revisions) (#98777)

engine-flutter-autoroll committed:
Roll Engine from e0d70c0bc2c0 to f6ea6186eecc (8 revisions) (#98784)

zanderso committed:
Revert "Reland "Enable caching of CPU samples collected at application startup (#89600)"" (#98803)

engine-flutter-autoroll committed:
Roll Plugins from 1e7af5e1d66b to d9fc677ab06f (4 revisions) (#98870)

engine-flutter-autoroll committed:
Roll Plugins from d9fc677ab06f to 6d88e3a2a184 (1 revision) (#98923)

justinmc committed:
Spacebar and enter in EditableText work with Inkwells (#98469)

Fixes a bug especially with text fields in a ListTile where space and enter didn't work.

tgucio committed:
Avoid unnecessary lambdas in SelectionOverlay.showHandles() (#98912)

bkonyi committed:
Update regexp for Observatory/VM Service message parsing for device lab (#98935)

Should unblock https://github.com/flutter/flutter/issues/98930

royarg02 committed:
Update comments for chip tests (#97476)

kirolous-nashaat committed:
Added optional parameter keyboardType to showDatePicker (#93439)

chunhtai committed:
Fix getOffsetForCaret to return correct value if contains widget span (#98542)

engine-flutter-autoroll committed:
Roll Engine from f6ea6186eecc to ef9d407fedfe (34 revisions) (#98940)

goderbauer committed:
Reland "Add more tests to slider to avoid future breakages (#98772)" (#98936)

gaaclarke committed:
Added a microbenchmark for Timeline events. (#98763)

tgucio committed:
Enable use_if_null_to_convert_nulls_to_bools lint (#98753)

chunhtai committed:
Remove redundant hide handles API from TextSelectionDelegate (#98944)

Commit:
Test that render object changed its visual appearance after texture is create (#98622)

himamis committed:
Add actionsOverflowAlignment parameter to dialog (#95995)

darrenaustin committed:
Fixed some issues and clarified the docs for ReorderableListView. (#98954)

Piinks committed:
Remove deprecated DayPicker and MonthPicker (#98543)

Swiftaxe committed:
Fixed order dependency and removed no-shuffle-tag in build_ios_framew… (#94699)

* Fixed order dependency and removed no-shuffle-tag in build_ios_framework_test.dart

* Removed trailing spaces in build_ios_framework_test.dart

* Removed unnecessary formatting

engine-flutter-autoroll committed:
Roll Engine from ef9d407fedfe to 258e7ff9c3df (6 revisions) (#98962)

werainkhatri committed:
Adds `onReorderStart` and `onReorderEnd` arguments to `ReorderableList`. (#96049)

engine-flutter-autoroll committed:
Roll Engine from 258e7ff9c3df to 1012ac4b5af7 (7 revisions) (#98976)

nt4f04uNd committed:
Correct the MediaQuery test name (#98984)

chunhtai committed:
Revert "Remove redundant hide handles API from TextSelectionDelegate (#98944)" (#99008)

This reverts commit 94f6fd6ea8934407452853405835b92fec0d7ad9.

chunhtai committed:
Clean up ClipboardStatusNotifier (#98951)

* Clean up ClipboardStatusNotifier

* fix

* fix test

* more refactaaaaagit add .

chingjun committed:
Add option in ProxiedDevice to only transfer the delta when deploying. (#97462)

Swiftaxe committed:
Removed no-shuffle tag and fixed order dependency in daemon_test.dart (#98970)

keyonghan committed:
Skip `can validate flutter version in parallel` test in `Linux web_tool_tests` (#99017)

tgucio committed:
Use int for PlaceholderSpan.placeholderCodeUnit (#98971)

xu-baolin committed:
Draggable can be accepted when the data is null (#97355)

dependabot[bot] committed:
Bump github/codeql-action from 1.1.2 to 1.1.3 (#99026)

jmagman committed:
Update Xcode version to 13a233 (#98297)

tgucio committed:
Call bringIntoView after RenderEditable updates on paste (#98604)

jmagman committed:
Remove cli_utils from allow-list (#99037)

Piinks committed:
Assert golden files use the right extension (#99016)

hterkelsen committed:
Revert "Assert golden files use the right extension (#99016)" (#99075)

This reverts commit ee86e61f73ea5a6880eef377febc3f2c37f1b2f6.

justinmc committed:
Pasting collapses the selection and puts it after the pasted content (#98679)

Desktop selection behavior improvement.

jmagman committed:
Increase recommended CocoaPods version to 1.11 (#98621)

brunotacca committed:
reads min/target sdk versions from localproperties (#98450)

DanTup committed:
[dap] Don't use --start-paused when running in Profile/Release mode (#98926)

gspencergoog committed:
Fix samples so that they present properly in the docs (#97224)

Commit:
Correct font weight for Cupertino tab label (#90109)

jmagman committed:
Wait for ios-deploy stdout before closing logLine stream (#99041)

werainkhatri committed:
adds `trackRadius` to `ScrollbarPainter` and `RawScrollbar` (#98018)

jonahwilliams committed:
use dart pub deps to analyze consumer dependencies (#99079)

xu-baolin committed:
Fix a Tabs crash when change the TabControllers (#98242)

fluttergithubbot committed:
Marks Linux_android android_picture_cache_complexity_scoring_perf__timeline_summary to be unflaky (#98989)

jonahwilliams committed:
[framework] Provide gesture settings to regular gesture detectors (#99081)

Piinks committed:
Re-land Assert golden files use the right extension (#99082)

xu-baolin committed:
Do not crash if dispatch the mouse events before the tooltip overlay detached (#97268)

TahaTesser committed:
[ReorderableListView] Add `footer` (#92086)

engine-flutter-autoroll committed:
Roll Plugins from 6d88e3a2a184 to 351415caa735 (8 revisions) (#99156)

TahaTesser committed:
Add `clipBehavior` to `Snackbar` (#98252)

keyonghan committed:
Deflake Mac module_test_ios (#99009)

jonahwilliams committed:
Use flutter pub instead of dart pub (#99159)

jmagman committed:
Run update-packages (#99038)

craiglabenz committed:
added wotw links for Baseline and GestureDetector (#99178)

christopherfujino committed:
[flutter_conductor] log to stderr but do not crash conductor if branch name validation fails (#99181)

engine-flutter-autoroll committed:
Roll Engine from 1012ac4b5af7 to 729f5b2ac016 (42 revisions) (#99190)

stuartmorgan committed:
Work around VS CMake generation bug (#98945)

engine-flutter-autoroll committed:
Roll Engine from 729f5b2ac016 to 5d190bb7f598 (1 revision) (#99193)

engine-flutter-autoroll committed:
Roll Engine from 5d190bb7f598 to 0726fdc4db38 (6 revisions) (#99213)

engine-flutter-autoroll committed:
Roll Engine from 0726fdc4db38 to f93dc3d7860f (1 revision) (#99216)

engine-flutter-autoroll committed:
Roll Engine from f93dc3d7860f to 97e2689b7b1d (7 revisions) (#99251)

engine-flutter-autoroll committed:
Roll Engine from 97e2689b7b1d to 05d1b5a66b8a (2 revisions) (#99258)

engine-flutter-autoroll committed:
Roll Engine from 05d1b5a66b8a to 4c00976e8602 (1 revision) (#99262)

engine-flutter-autoroll committed:
Roll Engine from 4c00976e8602 to efd538065a78 (1 revision) (#99267)

engine-flutter-autoroll committed:
Roll Plugins from 351415caa735 to f96d633e7a54 (3 revisions) (#99275)

fluttergithubbot committed:
Marks Mac_ios cubic_bezier_perf_ios_sksl_warmup__timeline_summary to be unflaky (#98991)

fluttergithubbot committed:
Marks Linux_android complex_layout__start_up to be flaky (#99000)

Commit:
Add regression test for #98973 (#99187)

jason-simmons committed:
Update an analyzer test to support changes to error message text in the latest Dart SDK (#99281)

See https://github.com/flutter/flutter/issues/99248

fluttergithubbot committed:
Marks Linux android views to be flaky (#99002)

jmagman committed:
Print events and views when first frame is taking awhile during tracing (#98957)

Commit:
Revert "reads min/target sdk versions from localproperties" (#99191)

engine-flutter-autoroll committed:
Roll Engine from efd538065a78 to 1a919ed90868 (1 revision) (#99278)

fluttergithubbot committed:
Marks Linux_android tiles_scroll_perf__timeline_summary to be unflaky (#98990)

albertodev01 committed:
feat: Added docstring examples to AnimatedBuilder and ChangeNotifier (#98628)

dkwingsmt committed:
[Keyboard] Correctly convert down events that are immediately synthesized released (#99200)

darrenaustin committed:
Updated to v0.86 of the Material Design tokens. (#99292)

Added the navigation_rail.json as well.

engine-flutter-autoroll committed:
Roll Engine from 1a919ed90868 to e3027b1b7bd4 (3 revisions) (#99289)

engine-flutter-autoroll committed:
Roll Engine from e3027b1b7bd4 to 615a60512c64 (2 revisions) (#99305)

Commit:
[cirrus] Refresh GCP service account (#99301)

GaryQian committed:
Use runUnsynchronized for android_views test (#99311)

engine-flutter-autoroll committed:
Roll Engine from 615a60512c64 to dc1dba505c33 (2 revisions) (#99310)

engine-flutter-autoroll committed:
Roll Plugins from f96d633e7a54 to 4eb17696ce2d (3 revisions) (#99342)

maheshj01 committed:
Add NavigationBar dartpad example  (#97046)

werainkhatri committed:
Migrates packages/flutter to from `hashValues` to `Object.hash` (#96109)

* migrates lib/foundation and lib/widgets

* migrates lib/painting and lib/material

* migrates lib/cupertino, lib/gestures, lib/services, lib/semantics and lib/rendering

* fixes failing tests

* migrates test/

* improves hashing of lists

Hari-07 committed:
Added viewport fraction parameter to tabView (#98512)

Commit:
[ci.yaml] Migrate roller to cocoon scheduler (#99162)

christopherfujino committed:
[flutter_conductor] deprecate increment (#99189)

jonahwilliams committed:
[framework] improve Notification API performance by skipping full Element tree traversal (#98451)

Renzo-Olivares committed:
Remove redundant properties passed to _Editable (#99192)

godofredoc committed:
Update mac chrome and driver to 98. (#99318)

engine-flutter-autoroll committed:
Roll Engine from dc1dba505c33 to a5718ab3c33b (9 revisions) (#99346)

* e97c8016e Started looking for the dart plugin registrant at runtime  (flutter/engine#31418)

* 9ebbc6eb0 Standardize the embedder file name. (flutter/engine#31723)

* e983ac898 Roll Skia from 1fc0c1c0f8af to 5deac304821b (2 revisions) (flutter/engine#31725)

* afd334d9d added dart plugin registrant test executables to run_tests.py (flutter/engine#31726)

* 0ff508b54 Standardize embedder archive name. (flutter/engine#31693)

* 9c1daae1e Roll Fuchsia Mac SDK from qGsRXn7-D... to RIOZP5nXz... (flutter/engine#31730)

* ceec638bb Roll Fuchsia Linux SDK from _mwyhkd9H... to 7rB0Yrprs... (flutter/engine#31731)

* 54050b405 Roll Skia from 5deac304821b to c3e7a2a00b0a (3 revisions) (flutter/engine#31732)

* a5718ab3c Roll Dart SDK from c3e3126a598b to 34a5563af80d (6 revisions) (flutter/engine#31734)

chunhtai committed:
Revert "Clean up ClipboardStatusNotifier (#98951)" (#99361)

This reverts commit c74a646b7bba1641d0b2b29f6024dc867844e94c.

Commit:
Add Android lifecycles test (#99319)

gaaclarke committed:
Starts using the `--source` flag to compile the dart registrant. (#98046)

* Starts using the `--source` flag to compile the dart registrant.

* updated general.shard tests

* Fixed the resident compiler flow

* added integration test

* made the integration test self contained

* renamed generated_main to dart_plugin_registrant

* cleaned up for review

* added task runner for ci

* added bringup and TESTOWNERS

* updated failure message

jmagman committed:
Update minimum required version to Xcode 13 (#97746)

Commit:
Revert "[ci.yaml] Migrate roller to cocoon scheduler (#99162)" (#99375)

markusaksli-nc committed:
Reland "Dismiss text selection toolbar with ESC" (#98995)

dependabot[bot] committed:
Bump actions/checkout from 2.4.0 to 3 (#99359)

engine-flutter-autoroll committed:
Roll Engine from a5718ab3c33b to 2c396fe8950f (15 revisions) (#99369)

engine-flutter-autoroll committed:
Roll Plugins from 4eb17696ce2d to 0283a99ce39d (8 revisions) (#99402)

xu-baolin committed:
Fix a Scrollbar hittest penetration issue (#99328)

Piinks committed:
Revert "Draggable can be accepted when the data is null" (#99419)

engine-flutter-autoroll committed:
Roll Engine from 2c396fe8950f to ac33354b28e0 (9 revisions) (#99424)

keyonghan committed:
Enable Samsung S10 tests in post-submit CI (#99353)

Piinks committed:
Limit overscroll stretching (#99364)

engine-flutter-autoroll committed:
Roll Engine from ac33354b28e0 to 45620d3de3fc (1 revision) (#99425)

fluttergithubbot committed:
Marks Windows_android complex_layout_win__compile to be unflaky (#99409)

joellurcook committed:
Simplify SafeArea test for maintainBottomViewPadding to ensure maintainBottomViewPadding is always respected (#97646)

tgucio committed:
Specify widget height in EditableText shortcuts tests (#98607)

goderbauer committed:
More beautiful linear_gradient sample (#99298)

engine-flutter-autoroll committed:
Roll Engine from 45620d3de3fc to d0ec2205b34e (2 revisions) (#99432)

Commit:
Fix lifecycle for API level 28 (#99433)

TahaTesser committed:
`CupertinoSliverNavigationBar`: Add example (#99384)

Part of #72926

gspencergoog committed:
Add material localizations for keyboard keys used for shortcut descriptions in menus. (#99020)

This change adds some localizations for keyboard keys that will be used in creating shortcut descriptions in menus.

engine-flutter-autoroll committed:
Roll Engine from d0ec2205b34e to 121376d39a56 (7 revisions) (#99446)

engine-flutter-autoroll committed:
Roll Engine from 121376d39a56 to 993285f64dac (6 revisions) (#99459)

zanderso committed:
Revert "Roll Engine from 121376d39a56 to 993285f64dac (6 revisions) (#99459)" (#99489)

This reverts commit a786f65700a1038494aa17c6bd7c0a0af0881ef8.

keyonghan committed:
Deflake post-submit Linux/Samsung S10 tests (#99454)

engine-flutter-autoroll committed:
Roll Engine from 121376d39a56 to a743fb81c763 (19 revisions) (#99501)

gspencergoog committed:
Fix the race condition in cache_test.dart (#99423)

This fixes a race condition in test/integration.shard/cache_test.dart which allowed the test to fail if the machine took a very long time to acquire a lock or launch a process.

Piinks committed:
Deprecate MaterialButtonWithIconMixin (#99088)

goderbauer committed:
Use `PlatformDispatcher.instance` over `window` where possible (#99496)

engine-flutter-autoroll committed:
Roll Engine from a743fb81c763 to 96200db8e706 (2 revisions) (#99509)

dependabot[bot] committed:
Bump actions/upload-artifact from 2.3.1 to 3 (#99515)

engine-flutter-autoroll committed:
Roll Engine from 96200db8e706 to 33afddeacd9b (5 revisions) (#99518)

Commit:
[ci.yaml] Mac web_tool_tests force cache clear via 98.1 (#99517)

engine-flutter-autoroll committed:
Roll Engine from 33afddeacd9b to 42814c3697df (4 revisions) (#99535)

engine-flutter-autoroll committed:
Roll Engine from 42814c3697df to 36197621dea9 (5 revisions) (#99537)

engine-flutter-autoroll committed:
Roll Engine from 36197621dea9 to 14b1f3fbc5e4 (3 revisions) (#99545)

engine-flutter-autoroll committed:
Roll Plugins from 0283a99ce39d to ab66a4bf3c4e (9 revisions) (#99556)

chunhtai committed:
Reland clean up clipboard. (#99363)

engine-flutter-autoroll committed:
Roll Engine from 14b1f3fbc5e4 to 9126849d86b1 (2 revisions) (#99560)

mbriand committed:
Fix custom-device default forwardPortSuccessRegex (#97719)

goderbauer committed:
Use `BindingBase.platformDispatcher` over `BindingBase.window` where possible (#99443)

christopherfujino committed:
fix type error in service_worker_test.dart (#99524)

TahaTesser committed:
Improve documentation of  `EditableText`/`TextField` callbacks (#98414)

werainkhatri committed:
complete migration of flutter repo to Object.hash* (#99505)

TahaTesser committed:
Migrate `Dialog` to Material 3 (#98919)

engine-flutter-autoroll committed:
Roll Engine from 9126849d86b1 to 2f819b16e6bb (2 revisions) (#99565)

r-mzy47 committed:
Improve container widget (#98389)

TahaTesser committed:
CupertinoButton: Add clickable cursor on web (#96863)

Commit:
Add Windows performance benchmark (#99564)

Commit:
[framework] add gesture settings to draggable (#99567)

engine-flutter-autoroll committed:
Roll Engine from 2f819b16e6bb to 42212138dfd0 (8 revisions) (#99571)

TahaTesser committed:
Make popup menu position configurable (#98979)

jmagman committed:
Add logging when first frame is not rendering (#99566)

jmagman committed:
Mark Mac_ios simple_animation_perf_ios unflaky (#99579)

engine-flutter-autoroll committed:
Roll Engine from 42212138dfd0 to 7b4905852d6f (3 revisions) (#99581)

godofredoc committed:
Add version to xcode name cache. (#99586)

Engine and Flutter using different xcode versions with the same cache
name is causing multiple problems when trying to find the runtimes. To
prevent this issue we will need to use different cache names for
different xcode versions.

Bug: https://github.com/flutter/flutter/issues/99429

engine-flutter-autoroll committed:
Roll Engine from 7b4905852d6f to e0e88d8c146f (1 revision) (#99594)

engine-flutter-autoroll committed:
Roll Engine from e0e88d8c146f to 6b5f8d13c380 (3 revisions) (#99609)

engine-flutter-autoroll committed:
Roll Engine from 6b5f8d13c380 to d0fe03eca4e5 (1 revision) (#99611)

skia-gold commented:
Gold has detected about 22 new digest(s) on patchset 8.
View them at https://flutter-gold.skia.org/cl/github/96598

Commit:
Add vs_build dep to windows performance test (#99602)

engine-flutter-autoroll committed:
Roll Engine from d0fe03eca4e5 to deace605c521 (1 revision) (#99620)

engine-flutter-autoroll committed:
Roll Engine from deace605c521 to b6b301e2d563 (2 revisions) (#99623)

engine-flutter-autoroll committed:
Roll Engine from b6b301e2d563 to e0648b42daaa (1 revision) (#99624)

engine-flutter-autoroll committed:
Roll Engine from e0648b42daaa to b2ba49443baf (1 revision) (#99630)

engine-flutter-autoroll committed:
Roll Engine from b2ba49443baf to 6576bd428ae1 (1 revision) (#99645)

engine-flutter-autoroll committed:
Roll Engine from 6576bd428ae1 to 1875f0ad3f54 (3 revisions) (#99649)

engine-flutter-autoroll committed:
Roll Engine from 1875f0ad3f54 to 97302b98fbb5 (1 revision) (#99657)

engine-flutter-autoroll committed:
Roll Engine from 97302b98fbb5 to 877bec343b35 (1 revision) (#99660)

engine-flutter-autoroll committed:
Roll Engine from 877bec343b35 to 898fadf19311 (1 revision) (#99666)

markusaksli-nc committed:
Re-invoke DismissIntent in Autocomplete if ignored (#99403)

skia-gold commented:
Gold has detected about 8 new digest(s) on patchset 13.
View them at https://flutter-gold.skia.org/cl/github/96598

engine-flutter-autoroll committed:
Roll Plugins from ab66a4bf3c4e to 675f91b3a3e2 (7 revisions) (#99687)

skia-gold commented:
Gold has detected about 35 new digest(s) on patchset 15.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 25 new digest(s) on patchset 16.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 23 new digest(s) on patchset 17.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 3 new digest(s) on patchset 19.
View them at https://flutter-gold.skia.org/cl/github/96598

engine-flutter-autoroll committed:
Roll Engine from 898fadf19311 to 2664593c3876 (2 revisions) (#99690)

skia-gold commented:
Gold has detected about 11 new digest(s) on patchset 21.
View them at https://flutter-gold.skia.org/cl/github/96598

Piinks commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,3647 @@
+
+    // Creates an element of randomness so that ink eminating from the same
+    // pixel have slightly different rings and sparkles.
+    _turbulenceSeed = math.Random().nextDouble() * 1000.0;
```
Oh, I see you intend to mock random below for the test. 👍

clocksmith replied:
Rather than mocking Random(), added a turbulence seed so that it can be constant, and a corresponding splashFactory. This is used for goldens, and can also be used if a developer wants a more deterministic effect.

I think mocking Random() would depend on mockito, but that doesn't appear to be a dep.

jonahwilliams committed:
Remove package:typed_data from package:flutter dependencies (#99604)

clocksmith commented on packages/flutter/test/material/ink_sparkle_test.dart:
```diff
@@ -0,0 +1,148 @@
+    final Offset bottomRight = tester.getBottomRight(buttonFinder);
+
+    final Offset centerTarget = topLeft + (bottomRight - topLeft) * 0.5;
+    await tester.tapAt(centerTarget);
```
Ah thanks! This makes the tests work now.

royarg02 committed:
[flutter_tools] Add timeout duration to error and handle exceptions for HttpHostValidator. (#98290)

skia-gold commented:
Gold has detected about 24 new digest(s) on patchset 23.
View them at https://flutter-gold.skia.org/cl/github/96598

Jasguerrero committed:
Support route on ios (#99078)

engine-flutter-autoroll committed:
Roll Engine from 2664593c3876 to f4e3f227d9aa (2 revisions) (#99696)

engine-flutter-autoroll committed:
Roll Engine from f4e3f227d9aa to 1d581abd3f09 (4 revisions) (#99705)

Piinks committed:
Remove deprecated RenderObjectElement methods (#98616)

TahaTesser committed:
CupertinoTabBar: Add clickable cursor on web (#96996)

engine-flutter-autoroll committed:
Roll Engine from 1d581abd3f09 to 9e1594bd741f (1 revision) (#99709)

dependabot[bot] committed:
Bump github/codeql-action from 1.1.3 to 1.1.4 (#99711)

christopherfujino committed:
correct github account usernames (#99716)

Piinks committed:
Remove deprecated Overflow and Stack.overflow (#98583)

dependabot[bot] committed:
Bump debian from bullseye-20220125-slim to bullseye-20220228-slim in /dev/ci/docker_linux (#99708)

engine-flutter-autoroll committed:
Roll Engine from 9e1594bd741f to 9c3f73864029 (2 revisions) (#99721)

skia-gold commented:
Gold has detected about 26 new digest(s) on patchset 24.
View them at https://flutter-gold.skia.org/cl/github/96598

clocksmith commented on packages/flutter/lib/src/material/ink_sparkle.dart:
```diff
@@ -0,0 +1,3647 @@
+  ///  1. If [customBorder] is provided, it is used to determine the path for
+  ///     clipping.
+  ///  2. If [customBorder] is null, and [borderRadius] is provided, the canvas
+  ///     is clipped by an [RRect] created from [_clipCallback] and
```
Changed to `_clipCallback`. https://github.com/flutter/flutter/issues/73163 will refactor some of these common methods and then public docs can be added.

Piinks committed:
Remove deprecated CupertinoTextField, TextField, TextFormField maxLengthEnforced (#98539)

skia-gold commented:
Gold has detected about 13 new digest(s) on patchset 26.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 25 new digest(s) on patchset 27.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 30 new digest(s) on patchset 30.
View them at https://flutter-gold.skia.org/cl/github/96598

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 32.
View them at https://flutter-gold.skia.org/cl/github/96598

clocksmith committed:
rebase

clocksmith commented:
moved to https://github.com/flutter/flutter/pull/99731
